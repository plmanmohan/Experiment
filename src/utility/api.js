import ApiConst from '../config/api-config.json'


/**
 * This class contains the generic method to call api
 * -returns the result of call
 */
export default function api() {
    return fetch( ApiConst.API, {
        method: ApiConst.API_METHOD,
        "headers": {
            "Content-Type": "application/json; charset=utf-8"
        }
    })
    .then (res => res.json())
    .catch(error => error)
}