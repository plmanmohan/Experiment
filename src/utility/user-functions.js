/**
 * It merges both of the arrays and removes the duplicate elements
 * @param {*} oldUserList
 * @param {*} newUserList
 */
export function removeDuplicateUsers(oldUserList, newUserList) {
    if (newUserList.length === 0) {
        return oldUserList;
    }
    let newData = oldUserList.concat(newUserList);

    let uniqueList = [];
    let finalData = [];

    for (let i = 0; i < newData.length; i++) {
        let fullName = newData[i].first_name + newData[i].last_name;
        if (uniqueList.indexOf(fullName) === -1) {
            uniqueList.push(fullName);
            finalData.push(newData[i]);
        }
    }

    console.log("AfterMerge :: ", finalData);
    return finalData;
}

/**
 * It search for the particular element with the id and removes the element from the Object
 * @param {*Object} state
 * @param {*number} id
 */
export function removeUser(users, id) {
    return users.filter(user => (user.id !== id ? user : null));
}
