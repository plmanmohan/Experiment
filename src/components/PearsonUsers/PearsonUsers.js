import React, { Component } from 'react';
import DisplayUser from './DisplayUser';
import sampleData from "../../config/data.json";
import api from '../../utility/api';
import { removeUser, removeDuplicateUsers } from "../../utility/user-functions";
import ConfirmationBox from '../../utility/ConfirmationBox/ConfirmationBox';
import "./style.css";

export default class PearsonUsers extends Component {

    constructor() {
        super();
        this.state = {
            users: [...sampleData],
            isLoading: true,
            error: false,
            showConfirmationBox: false,
            currentId: null
        }
    }
    /**
     * This method is called to fetch data 
     */
    componentDidMount() {
        this.fetchApiData();
    }

    /**
     * Here api call is done within this method
     * - then the data is merged into previous data and duplicate isers are removed
     */
    fetchApiData = () => {
        api()
            .then(result => {
                const updatedUserList = removeDuplicateUsers(this.state.users, result.data || sampleData);
                const error = result.error || false
                this.setState({
                    users: [...updatedUserList],
                    isLoading: false,
                    error
                })
            })
            .catch(() => {
                this.setState({ error: true, isLoading: false });
            })
    }

    /**
     * This method is used to hide errorbox
     */
    disableErrorBox = (val) => this.setState({ error: false });

    /**
     * This method is called when any of the users is selected.
     */
    selectUserToRemove = (id) => {
        this.setState({ showConfirmationBox: true, currentId: id })
    }
    /**
     * This method defines if the selected user is to be removed or not, based on the confirmation 
     * okay / cancel button click.
     */
    removeUser = (val) => {
        if (val) {
            this.setState({ users: removeUser(this.state.users, this.state.currentId), currentId: null, showConfirmationBox: false })
        } else {
            this.setState({ currentId: null, showConfirmationBox: false })
        }
    }
    /**
     * This method is used to create the DisplayUser components based on the user data
     */
    createDisplayUsers = () => {
        return this.state.users.map(user => <DisplayUser key={user.id} onClick={this.selectUserToRemove} data={user} />)
    }
    
    render() {
        const { isLoading, error } = this.state;
        return (
            <div className="pearon-users">
                <h1 className="heading">Pearson User Management </h1>
                <div className="user-container">
                    {isLoading ? (
                        <div className="loader">
                            <img
                                height={80}
                                width={80}
                                src="https://upload.wikimedia.org/wikipedia/commons/3/3a/Gray_circles_rotate.gif"
                                alt="loading"
                            />
                        </div>
                    ) :
                        this.createDisplayUsers()
                    }
                </div>
                {error && <ConfirmationBox onClick={this.disableErrorBox} message={error.message || "Error in Loading!"} isAlert={true} />}
                {this.state.showConfirmationBox ? <ConfirmationBox onClick={this.removeUser} message="Are you sure you want to delete the user!" isAlert={false} /> : null}
            </div>
        );
    }
}
