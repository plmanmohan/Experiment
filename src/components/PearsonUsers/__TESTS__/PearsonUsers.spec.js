import  PearsonUsers  from "../PearsonUsers";
import { removeUser, removeDuplicateUsers } from "../../../utility/user-functions";

//mock test for api
jest.mock('../../../utility/api');

describe('Test PearsonUsers component', () => {

  it('renders PearsonUsers without crashing', () => {
    shallow(<PearsonUsers />);
  });

  it('api() call test', () => {
    return api().then(data => {
      expect(data).toStrictEqual({
        "data":{ 
                id: 4,
                first_name: "Eve",
                last_name: "Holt",
                avatar: "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        }
      });
    });
  });

  it("Render Heading text test", () => {
    const wrapper = shallow(<PearsonUsers />);
    expect(wrapper).toMatchSnapshot();
    const h1 = wrapper.find("h1");
    expect(h1.text()).toEqual("Pearson User Management ");
  });

  //remove duplicated users from the state add only unique into state
  it("removeDuplicateUsers removes duplicates adds unique users into state and ", () => {
    const oldUserList = [
      { id: 2, first_name: "Janet", last_name: "Weaver", avatar: "128.jpg" },
      { id: 3, first_name: "Emma", last_name: "Wong", avatar: "128.jpg" }
    ];
    const newUserList = [
      { id: 12, first_name: "Rachel", last_name: "Howell", avatar: "128.jpg" },
      { id: 2, first_name: "Janet", last_name: "Weaver", avatar: "128.jpg" }
    ];
    const finalUserList = removeDuplicateUsers(oldUserList, newUserList);
    expect(finalUserList).toEqual([
      { id: 2, first_name: "Janet", last_name: "Weaver", avatar: "128.jpg" },
      { id: 3, first_name: "Emma", last_name: "Wong", avatar: "128.jpg" },
      { id: 12, first_name: "Rachel", last_name: "Howell", avatar: "128.jpg" }
    ]);
  });

  //removeUser method is tested
  it("removeUser removes the user", () => {
    const state = {
      users: [                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
        {
          id: 2,
          first_name: "Janet",
          last_name: "Weaver",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg"
        },
        {
          id: 3,
          first_name: "Emma",
          last_name: "Wong",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/olegpogodaev/128.jpg"
        },
        {
          id: 4,
          first_name: "Eve",
          last_name: "Holt",
          avatar:
            "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
        }
      ]
    };
    const id = 3;
    const finalState = removeUser(state.users, id);
    expect(finalState).toEqual([
      {
        id: 2,
        first_name: "Janet",
        last_name: "Weaver",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg"
      },
      {
        id: 4,
        first_name: "Eve",
        last_name: "Holt",
        avatar:
          "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg"
      }
    ]);
  });
  

  //disableErrorBox
  it("disableErrorBox is tested", () => {
    const wrapper = shallow(<PearsonUsers />);
    wrapper.setState({ error: true }, () => {
      wrapper.instance().disableErrorBox(true);
      const userState = wrapper.state('error');
      expect(userState).toEqual(false);
    })
  });

  //selectUserToRemove
  it("selectUserToRemove is tested", () => {
    const wrapper = shallow(<PearsonUsers />);
    wrapper.setState({ showConfirmationBox: false, currentId: -1 }, () => {
      wrapper.instance().selectUserToRemove(2);
      const currentUserId = wrapper.state('currentId');
      expect(currentUserId).toEqual(2);
    })
  });

  //removeUser
  it("removeUser is tested", () => {
    const wrapper = shallow(<PearsonUsers />);
    wrapper.setState({ showConfirmationBox: true, currentId: 2 }, () => {
      wrapper.instance().removeUser(true);
      const currentUserId = wrapper.state('currentId');
      expect(currentUserId).toEqual(null);
    })
  });
});

