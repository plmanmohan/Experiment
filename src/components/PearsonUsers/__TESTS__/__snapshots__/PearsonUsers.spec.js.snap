// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PearsonUsers component Render Heading text test 1`] = `ShallowWrapper {}`;
